include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,621 @@
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/sundials/kinsol.h>
+
+#include <deal.II/trilinos/nox.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * Selects a nonlinear solver by changing a parameter.
+ *
+ * By calling the @p solve function of this @p NonlinearSolverSelector, it selects the
+ * @p solve function of that @p Solver that was specified in the constructor
+ * of this class, similar to the @p SolverSelector class.
+ *
+ * <h3>Usage</h3> An example of code one would run with this class is the
+ * following:
+ * @code
+ * // Generate a @p NonlinearSolverSelector that uses @p KINSOL
+ * std::string solver = "kinsol";
+ *
+ * NonlinearSolverSelector<Vector<double>>::AdditionalData additional_data;
+ * additional_data.solver_type = solver;
+ *
+ * NonlinearSolverSelector<Vector<double>> nonlinear_solver(additional_data);
+ *
+ * // Functions that are required for solving a nonlinear problem,
+ * // which are utilized in both @p KINSOL and @p NOX.
+ * nonlinear_solver.reinit_vector = [&](Vector<double> &x) {...}
+ *
+ * nonlinear_solver.residual =
+ *           [&](const Vector<double> &evaluation_point,
+ *               Vector<double> &      residual) {...}
+ *
+ * nonlinear_solver.setup_jacobian =
+ *           [&](const Vector<double> &current_u,
+ *               const Vector<double> &current_f) {...}
+ *
+ * nonlinear_solver.solve_with_jacobian =
+ *           [&](const Vector<double> &rhs,
+ *               Vector<double> &      dst,
+ *               const double tolerance) {...}
+ *
+ * // Calling the @p solve function with an initial guess.
+ * nonlinear_solver.solve(current_solution);
+ * @endcode
+ */
+template <typename VectorType = Vector<double>>
+class NonlinearSolverSelector
+{
+public:
+  /**
+   * Additional data that will be sent through the NonlinearSolverSelector
+   * class, then into the specific nonlinear solver classes themselves.
+   * Many of these parameters can also be foud in the documentation for
+   * @p KINSOL and @p NOX.
+   */
+  class AdditionalData
+  {
+  public:
+    /**
+     * NonlinearSolverSelector solution strategy. The solver types included in
+     * this class both use a Newton-Krylov solver with a line search and even
+     * Picard iterations (KINSOL).
+     */
+    enum SolutionStrategy
+    {
+      /**
+       * Standard Newton iteration.
+       */
+      newton,
+      /**
+       * Newton iteration with linesearch.
+       */
+      linesearch,
+      /**
+       * Picard iteration.
+       */
+      picard,
+    };
+
+    /**
+     * Initialization parameters for NonlinearSolverSelector.
+     *
+     * Global parameters:
+     *
+     * @param solver_type nonlinear solver type
+     * @param maximum_non_linear_iterations Maximum number of nonlinear
+     * iterations
```

Comment:
Here and below, end each sentence with a period:
```suggestion
     * @param maximum_non_linear_iterations Maximum number of nonlinear
     * iterations.
```